File: 792_Number_of_Matching_Subsequences.cpp
class Solution {
public:
    int numMatchingSubseq(string S, vector<string>& words) {
        vector<vector<int> > alpha(26);
        int result = 0;
        for (int i = 0; i < S.size(); i++)
            alpha[S[i] - 'a'].push_back(i);
        for (auto word : words) {
            bool found = true;
            int index = -1;
            for (auto c : word) {
                auto it = upper_bound(alpha[c - 'a'].begin(), alpha[c - 'a'].end(), index);
                if (it >= alpha[c-'a'].end()) found = false;
                else index = *it;
            }
            if (found) result++;
        }
        return result;
    }
};

File: 792_Number_of_Matching_Subsequences.py
from collections import defaultdict
from bisect import bisect_right
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        ret = 0
        lookup = defaultdict(list)
        for i, c in enumerate(s):
            lookup[c].append(i)
        for word in words:
            cur, found = -1, True
            for w in word:
                idx = bisect_right(lookup[w], cur)
                if idx >= len(lookup[w]):
                    found = False
                    break
                else:
                    cur = lookup[w][idx]
            if found:
                ret += 1
        return ret
